### DISCUSSION TITLE
export CUE definition into a Go structure

### DISCUSSION THREAD

**Discussion by owulveryck** [cuelang/cue#1038] [Q&A]
Hi,

I have a definition file that looks like this:

```cue
package mypackage

#Infrastructure: {
	// the Kafka topic where the information should be sent
	topic: string
	// the endpoint from where the information can be pulled
	endpoint: string
}

// Definition of the envelop
#MessageEnvelop: {
	subject?:         =~"^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$" // a uuid
	source:           =~"^bla.+"
	type?:            =~"^bla.+:[cud]" // c : create u : update d : delete
	consent?:         string
	dataContentType?: string
}

#Configuration: {
	infrastructure: #Infrastructure
	messageEnvelop: #MessageEnvelop
	dataVersion:    string
}
```

I'd like to generate the corresponding Go structures; I expect something like:

```go
type Infrastructure struct {
   Topic string
   Endpoint string
 }
// ...
```

I tried to run 
```
❯ cue export --out go schemas/event/event.cue
unsupported encoding "code"
```

Is it the idiomatic way to create a go schema?
Why does it fail?

**Comment by myitcv** [comment]
Please see https://cuelang.org/docs/integrations/go/#generate-go-code (note that as a caveat for now you will end up using deprecated identifiers: this will be fixed by https://github.com/cuelang/cue/issues/1035)

**Comment by verdverm** [comment]
In general, CUE doesn't have the ability to generate the structs in a target language. There are several past issues and discussions you might search for here. We've been using `text/template` to achieve some results. @myitcv references the ability to generate code which validates an object against the schema defined in CUE.

Some interesting difficulties are:

- CUE has different numeric types from most languages, how do they map to a given language? What about lists and maps?
- Public vs private visibility, how does one define optional in language X?
- Differing package and import denotation / semantics across languages
- differing styles between languages and companies, how would one provide flexibility to conform to different lint preferences for (in example) JS

There may also be ideas around proposed solutions to the problem. I've personally written some generators for Go structs with https://github.com/hofstadter-io/hof

**Reply by myitcv** [reply]
@verdverm thanks for adding additional clarification. I totally forgot to reference https://github.com/cuelang/cue/issues/6 in my response above; i.e. it's an explicit goal of CUE to be able to "convert" to a programming language encoding just as it is to/from protobuf, JSON schema etc. As you point out there may well be limitations of such a conversion, but it remains a goal nonetheless.

**Comment by cueckoo** [comment]
This discussion has been migrated to https://github.com/cue-lang/cue/issues/1038.

For more details about CUE's migration to a new home, please see https://github.com/cue-lang/cue/issues/1078.
